include/nbl/video/CThreadSafeGPUQueueAdapter.h

```diff
@@ -0,0 +1,40 @@
+#ifndef __IRR_GPU_QUEUE_THREADSAFE_ADAPTER_H_INCLUDED__
+#define __IRR_GPU_QUEUE_THREADSAFE_ADAPTER_H_INCLUDED__
+#include "IGPUQueue.h"
+
+namespace nbl::video
+{
+    /*
+        A thread-safe implementation of IGPUQueue.
+        Note, that using the same queue as both a threadsafe queue and a normal queue invalidates the safety.
+    */
+    class CThreadSafeGPUQueueAdapter : public IGPUQueue
+    {
+    protected:
+        core::smart_refctd_ptr<IGPUQueue> originalQueue = nullptr;
+        std::mutex m;
+    public:
+        CThreadSafeGPUQueueAdapter(nbl::core::smart_refctd_ptr<IGPUQueue>&& original) : IGPUQueue(nullptr, original->getFamilyIndex(), original->getFlags(), original->getPriority()),
```

Comment:
`original->getOriginDevice()` instead of nullptr
All GPU-side objects need to know which logical device they belong to

Reply:
Yeah i think i didn't add that cause it was a protected function

Reply:
oh, right, then just take device pointer as ctor param. Make sure `original` comes from the same device.
Or make getOriginDevice public, your call